src/clib/pio_darray.c

```diff
@@ -1694,19 +1818,37 @@ static int PIOc_write_darray_adios(file_desc_t *file, int varid, int ioid,
     int can_merge_buffers = 1;
     if (file->block_myrank == 0)
     {
-        if ((av->elem_size*buffer_count) >= file->block_array_size)
+        if ((av->adios_type_size * buffer_count) > file->block_array_size)
         {
-            can_merge_buffers = 0;
+            file->block_array_size = (size_t)(av->adios_type_size * buffer_count);
+            file->block_array = (char*)realloc(file->block_array, file->block_array_size);
+            if (file->block_array == NULL)
+            {
+                file->block_array_size = 0;
+                can_merge_buffers = 0;
+            }
         }
     }
+
     MPI_Bcast(&can_merge_buffers, 1, MPI_INT, 0, file->block_comm);
+    if (can_merge_buffers != 1)
+    {
+        return pio_err(NULL, file, PIO_ENOMEM, __FILE__, __LINE__,
+                       "Could not allocate memory to merge buffers fo variable (decomp_id/%s) failed (adios2_error=%s) for file (%s, ncid=%d)",
+                       av->name, convert_adios2_error_to_string(adiosErr), pio_get_fname_from_file(file), file->pio_ncid);
+    }
 
     if (can_merge_buffers == 1)
```

Comment:
Since the function returns when (can_merge_buffers != 1), do we still need to check the value here?

Reply:
No. That check can be removed. It was there for debugging purposes, to quickly switch between no-merge and merge implementations.